scripts/check_backend_associated_test_file.py

```diff
@@ -0,0 +1,134 @@
+    'scripts/linters/test_files/invalid_copyright.py',
+    'scripts/linters/test_files/invalid_docstring.py',
+    'scripts/linters/test_files/invalid_import_order.py',
+    'scripts/linters/test_files/invalid_merge_conflict.py',
+    'scripts/linters/test_files/invalid_ndb.py',
+    'scripts/linters/test_files/invalid_no_newline.py',
+    'scripts/linters/test_files/invalid_pycodestyle_error.py',
+    'scripts/linters/test_files/invalid_pylint_id.py',
+    'scripts/linters/test_files/invalid_python_three.py',
+    'scripts/linters/test_files/invalid_request.py',
+    'scripts/linters/test_files/invalid_tabs.py',
+    'scripts/linters/test_files/invalid_todo.py',
+    'scripts/linters/test_files/invalid_urlopen.py',
+    'scripts/linters/test_files/valid.py',
+    'scripts/linters/test_files/valid_py_ignore_pragma.py',
+    'core/tests/build_sources/extensions/CodeRepl.py',
+    'core/tests/build_sources/extensions/DragAndDropSortInput.py',
+    'core/tests/data/failing_tests.py',
+    'core/tests/data/image_constants.py',
+    'core/tests/data/unicode_and_str_handler.py',
+    'proto_files/text_classifier_pb2.py',
+    'proto_files/training_job_response_payload_pb2.py',
+
+    # The files below are to be tested.
+    # GSoC'22: Improve line and branch coverage for the backend and frontend.
+    'scripts/run_frontend_tests.py',
+    'scripts/create_expression_parser.py',
+    'scripts/run_custom_eslint_tests.py',
+    'scripts/run_backend_tests.py',
+    'scripts/run_lighthouse_tests.py',
+    'scripts/run_portserver.py',
+    'scripts/run_presubmit_checks.py',
+    'scripts/run_tests.py',
+    'scripts/start.py',
+    'setup.py',
+    'scripts/check_backend_associated_test_file.py',
+    'scripts/third_party_size_check.py',
+    'scripts/linters/warranted_angular_security_bypasses.py',
+    'core/feconf.py',
+    'core/domain/cron_services.py',
+    'core/platform/datastore/cloud_datastore_services.py',
+    'core/platform/transactions/cloud_transaction_services.py',
+]
+
+
+def main() -> None:
+    """Finds the non-empty backend files that lack an associated test file."""
+
+    all_backend_files = []
+    for (root, _, files) in os.walk(TOPMOST_LEVEL_PATH):
+        for file in files:
+            full_path = os.path.join(root, file)
+            _, file_extension = os.path.splitext(full_path)
+            if (file_extension == '.py') and (
+                '/third_party/' not in full_path) and (
+                    '/node_modules/' not in full_path):
+                all_backend_files.append(full_path)
+
+    all_backend_files.sort()
+    files_without_test = []
+    for file in all_backend_files:
+        if not file.endswith('_test.py'):
+            # Replace last occurance of '.py' with '_test.py' to get the
+            # associated test file.
+            to_be_replaced = '.py'
+            to_replace_with = '_test.py'
+            test_file = file[::-1].replace(
+                to_be_replaced[::-1], to_replace_with[::-1], 1)[::-1]
```

Comment:
This is confusing why do we need so many `[::-1]`

Reply:
We need to replace only the last occurrence of `.py` with `_test.py`. 
The `replace()` method in Python accepts three parameters: old string, new string and the number of replacements to be made from the start. [See documentation](https://docs.python.org/3/library/stdtypes.html#str.replace).
What I am doing here, is that I am reversing the string and then replacing the first occurrence of `.py` (actually 'yp.' in the reversed string) with `_test.py` (actually 'yp.tset_' in the reversed string). I am making sure that only the first occurance is replaced in the reversed string by passing the value of count parameter as 1.

Reply:
I have added a different suggestion to simplify this.